c/tskit/trees.c

```diff
@@ -6803,3 +6803,280 @@ tsk_treeseq_divergence_matrix(const tsk_treeseq_t *self, tsk_size_t num_samples,
 out:
     return ret;
 }
+
+/* ======================================================== *
+ * Extend edges
+ * ======================================================== */
+
+typedef struct _edge_list_t {
+    tsk_id_t edge;
+    // the `extended` flags records whether we have decided to extend
+    // this entry to the current tree?
+    bool extended;
+    struct _edge_list_t *next;
+} edge_list_t;
+
+static int
+extend_edges_append_entry(
+    edge_list_t **head, edge_list_t **tail, tsk_blkalloc_t *heap, tsk_id_t edge)
+{
+    int ret = 0;
+    edge_list_t *x = NULL;
+
+    x = tsk_blkalloc_get(heap, sizeof(*x));
+    if (x == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+
+    x->edge = edge;
+    x->extended = false;
+    x->next = NULL;
+
+    if (*tail == NULL) {
+        *head = x;
+    } else {
+        (*tail)->next = x;
+    }
+    *tail = x;
+out:
+    return ret;
+}
+
+static void
+remove_unextended(edge_list_t **head, edge_list_t **tail)
+{
+    edge_list_t *px, *x;
+
+    px = *head;
+    while (px != NULL && !px->extended) {
+        px = px->next;
+    }
+    *head = px;
+    if (px != NULL) {
+        px->extended = false;
+        x = px->next;
+        while (x != NULL) {
+            if (x->extended) {
+                x->extended = false;
+                px->next = x;
+                px = x;
+            }
+            x = x->next;
+        }
+    }
+    *tail = px;
+}
+
+static int
+tsk_treeseq_extend_edges_iter(
+    const tsk_treeseq_t *self, int direction, tsk_edge_table_t *edges)
+{
+    // Note: this modifies the edge table, but it does this by (a) removing
+    // some edges, and (b) extending left/right endpoints of others,
+    // while keeping order the same, and so this maintains sortedness
+    // (so, there is no need to sort afterwards).
+    int ret = 0;
+    tsk_id_t tj;
+    tsk_id_t e, e1, e2, e_in;
+    tsk_blkalloc_t edge_list_heap;
+    double *near_side, *far_side;
+    edge_list_t *edges_in_head, *edges_in_tail;
```

Comment:
If I was starting this from scratch, I would't bother with linked lists and do something like this instead:
```
tsk_bool_t extended = calloc(num_edges, sizeof(*extended));
tsk_id_t *edges_in = mallco(num_edges, sizeof(*edges_in));
tsk_id_t *edges_out = mallco(num_edges, sizeof(*edges_out));
tsk_size_t num_edge_in, num_edges_out;
```

and keep track of the lists like that. Repacking to get rid of unextended edges will be slightly trickier, but there's less faddling around with linked lists, memory allocation and so on, and it would almost certainly be more efficient.

But like I say - you might not want to get into this, since it's working.

and keep track of th